### PR TITLE
Add fastpath to isRelatedTo for type references

### LINKED ISSUE #37344: Recursive type alias in union type breaks in 3.8.3
**TypeScript Version:**  3.8.3

**Search Terms:** recursive, recursion

**Code**

```ts
type Action<T, P> = P extends void ? { type : T } : { type: T, payload: P }

enum ActionType {
    Foo,
    Bar,
    Baz,
    Batch
}

type ReducerAction =
  | Action<ActionType.Bar, number>
  | Action<ActionType.Baz, boolean>
  | Action<ActionType.Foo, string>
  | Action<ActionType.Batch, ReducerAction[]>
```

**Expected behavior:** compiler does not complain recursive type alias, works fine in 3.7.5

**Actual behavior:** compiler complains recursive type alias, breaks in 3.8.3
`Type arguments for 'Array' circularly reference themselves.`

**Playground Link:** 
http://www.typescriptlang.org/play/index.html#code/C4TwDgpgBAggxsAlgewHYB4AqAaKAFAPigF58oIAPYCVAEwGcoA3ZRWqAfigG8pRIoALiiYoAXyE8+4CMJxQwAQxAAbZItrC84gFA6aAVwC2sBClSYZPHVFtQAYsmTYbdgEKKATi7tQPALx93RWA4AAsdMT1+aAAlCFoDOAhPeCQ0ElcAH1N0jDTzS0gAOg9vKFRjACMUgmzc83QCtCKIUsVAqCqnFQhFVDrbHOb8sxaZYsdnKHpgT0RUAHNBqGGx0bzW9tCw3HjE5NT1gG0AXTqgA

**Related Issues:**
I'm not sure, if this might be related to https://github.com/microsoft/TypeScript/issues/35017.

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #37344

This is an alternative to #37423. Rather than deferring conditional simplification, we introduce a fastpath for relating type references to primitives, which occurs _before_ normalization, this way we don't pull on the type arguments for deferred type references unless we're much more likely to actually need their structure to resolve the relationship.

### CODE REVIEW

**Comment by weswigham** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -15340,11 +15340,23 @@ namespace ts {
...
              */
             function isRelatedTo(originalSource: Type, originalTarget: Type, reportErrors = false, headMessage?: DiagnosticMessage, intersectionState = IntersectionState.None): Ternary {
+                // Before normalization: if `source` is type reference (ergo an object), and `target` is primitive,
+                // skip all the checks we don't need and just return `isSimpleTypeRelatedTo` result
+                if (getObjectFlags(originalSource) & ObjectFlags.Reference && originalTarget.flags & TypeFlags.Primitive) {
+                    if (isSimpleTypeRelatedTo(originalSource, originalTarget, relation, reportErrors ? reportError : undefined)) {
+                        return Ternary.True;
+                    }
+                    else {
+                        reportErrorResults(originalSource, originalTarget, Ternary.False, /*isComparingJsxAttributes*/ true);
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `/*isComparingJsxAttributes*/ true`?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -15340,11 +15340,23 @@ namespace ts {
              * * Ternary.False if they are not related.
              */
             function isRelatedTo(originalSource: Type, originalTarget: Type, reportErrors = false, headMessage?: DiagnosticMessage, intersectionState = IntersectionState.None): Ternary {
+                // Before normalization: if `source` is type reference (ergo an object), and `target` is primitive,
+                // skip all the checks we don't need and just return `isSimpleTypeRelatedTo` result
+                if (getObjectFlags(originalSource) & ObjectFlags.Reference && originalTarget.flags & TypeFlags.Primitive) {
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts]: `getObjectFlags` on an intersection will bubble up `ObjectFlags.Reference`. That means that a primitive branded with a type reference (maybe rare) will no longer be compatible, right?

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

**Comment by DanielRosenwasser** [comment]
Can't seem to find behavior that exhibited the case I was mentioning above. I tried

```ts
interface Foo<T> {
    yadda: T;
}

declare let x: number & Foo<string>;
declare let y: number;

x = y;
y = x;
```

but that doesn't seem to have any differences.

**Comment by DanielRosenwasser** [comment]
Are type references special here? Can you just check if the source has `TypeFlags.Object` too?

**Comment by weswigham** [comment]
> but that doesn't seem to have any differences.

That's because the only type flags propagated are the propagating flags `ContainsWideningType | ContainsObjectOrArrayLiteral | NonInferrableType`.

> Are type references special here? Can you just check if the source has TypeFlags.Object too?

Maybe, I'll simplify it.

**Comment by DanielRosenwasser** [comment]
Sorry, I misread which flags were getting propagated.

**Comment by DanielRosenwasser** [comment]
@typescript-bot user test this

**Comment by weswigham** [comment]
@typescript-bot perf test this
@typescript-bot user test this
@typescript-bot test this
@typescript-bot run dt

**Comment by weswigham** [comment]
@typescript-bot pack this

**Comment by weswigham** [comment]
@ahejlsberg can I get a review?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -15541,11 +15541,23 @@ namespace ts {
              * * Ternary.False if they are not related.
              */
             function isRelatedTo(originalSource: Type, originalTarget: Type, reportErrors = false, headMessage?: DiagnosticMessage, intersectionState = IntersectionState.None): Ternary {
+                // Before normalization: if `source` is type an object type, and `target` is primitive,
+                // skip all the checks we don't need and just return `isSimpleTypeRelatedTo` result
+                if (originalSource.flags & TypeFlags.Object && originalTarget.flags & TypeFlags.Primitive) {
+                    if (isSimpleTypeRelatedTo(originalSource, originalTarget, relation, reportErrors ? reportError : undefined)) {
+                        return Ternary.True;
+                    }
+                    else {
```
Comment (ahejlsberg) [inline src/compiler/checker.ts:15550]: No need for `else` block here as the true case always returns.

**Pull request merged**